src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java

```diff
@@ -27,16 +35,168 @@
     {
         super(entity);
-        this.target = target;
+        this.targetType = LivingTargetType.GOAL_TARGET;        
+        this.originalTarget = target;
+        
+        assertLivingIsValid();
+    }
+    
+    public LivingSetAttackTargetEvent(LivingEntity entity, LivingEntity target, ILivingTargetType targetType)
+    {
+        super(entity);
+        this.targetType = targetType;
+        this.originalTarget = target;
+        
+        assertLivingIsValid();
     }
 
+    /**
+     * {@return the target this living entity had after changing its target to a new
+     * target, but before posting this event. The return value of this method cannot be
+     * affected by {@link LivingSetAttackTargetEvent#setTarget}}
+     */
     public LivingEntity getTarget()
     {
-        return target;
+        return originalTarget;
+    }
+    
+    /**
+     * {@return the current target this living entity has. The return value of this
+     * method may be affected by {@link LivingSetAttackTargetEvent#setTarget}}
+     */
+    public Optional<LivingEntity> getCurrentTarget()
+    {
+        return targetType.getTarget(getEntity());
+    }
+
+    /**
+     * @param target The living entity that shall be set as a target.
+     * {@return {@code true} when the target is valid, otherwise {@code false}.}
+     */
+    public boolean setTarget(@Nullable LivingEntity target)
+    {
+        if(targetType.isValidTarget(target))
+        {
+            targetType.setTarget(getEntity(), target);
+            return true;
+        }
+        
+        return false;
+    }
+    
+    /**
+     * {@return the target type of this event.}
+     */
+    public ILivingTargetType getTargetType()
+    {
+        return targetType;
+    }
+    
+    /**
+     * Throws an exception when the living entity is not valid for this target type.
+     */
+    private void assertLivingIsValid()
+    {
+        if(!this.targetType.isValidLiving(getEntity()))
+        {
+            throw new IllegalArgumentException("The living entity provided to this event is not compatible with the chosen target type.");
+        }
+    }
+    
+    // Short note on why this is needed: If we were to just let the event "return" data like in the ComputeFovModifierEvent,
+    // this would introduce lots of bugs and breaking changes, so we need to do this mess.
+    // 
+    // Note for any future developer that might stumble across this: Please introduce a breaking change
+    // later down the line (like, when porting forge to 1.20 or 1.21; this is the reaons why all the members of this class are marked for removal) 
+    // and do all of this like it is done in the ComputeFovModifierEvent. By that time, most mod developers will
+    // (hopefully) use LivingSetAttackTargetEvent#setTarget instead of setting the target on their own, which will
+    // minimize the impact of such a breaking change.
+    public static interface ILivingTargetType
```

Comment:
Why does this need to be public and therefor be subject to breaking changes? I think this interface should just be gone, the enum made private and used directly internally.
To then allow creating the event with the two different types you can introduce static factory methods.

Reply:
The reason that I used an interface is to make the event extensible for other mods (that maybe add new ways of setting a Target). However, after thinking about it a bit, I must admit that this will probably never happen, so, yeah, I should probably remove the interface.